Allow creating release source code archives with make release

wentasah opened the pull request:
I tried to create the archives the same way as at
https://github.com/darrylb123/usbrelay/releases/tag/0.7 . I'm not sure
whether these were created manually or Github allows creating them
automatically. In any way, it should be possible to extend the
Makefile even more so that `make release` uploads the created archives
to GitHub. Darryl, let me know if you're interested in this.

darrylb123 commented:
Interesting. I hadn't seen them so I didn't create them. Must be done by
github when the tag is created. I'm never sure when to create new tags.
Perhaps I should created more of them and call the stable ones that (or
release). Then document how to get the stable ones and link. Doesn't help
with git version.h though. I wonder if there is any control over the naming
so that the archive is called the git version, then use that in the
makefile.

Darryl

On Mon., 18 May 2020, 7:01 pm Michal Sojka, <notifications@github.com>
wrote:

> I tried to create the archives the same way as at
> https://github.com/darrylb123/usbrelay/releases/tag/0.7 . I'm not sure
> whether these were created manually or Github allows creating them
> automatically. In any way, it should be possible to extend the
> Makefile even more so that make release uploads the created archives
> to GitHub. Darryl, let me know if you're interested in this.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/darrylb123/usbrelay/pull/55
> Commit Summary
>
>    - Allow creating release source code archives with make release
>
> File Changes
>
>    - *M* Makefile
>    <https://github.com/darrylb123/usbrelay/pull/55/files#diff-b67911656ef5d18c4ae36cb6741b7965>
>    (24)
>
> Patch Links:
>
>    - https://github.com/darrylb123/usbrelay/pull/55.patch
>    - https://github.com/darrylb123/usbrelay/pull/55.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/darrylb123/usbrelay/pull/55>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTSUVLBL5IPFUFVIYJIK7TRSD2N3ANCNFSM4ND4GOCA>
> .
>

darrylb123 commented:
Just had a look at the release tar,gz file. The top level directory is the release version.
Perhaps that is what we should use. 
- Call the releases a standard name like continue using the numerical name or use the git hash. So the build directory would be usbrelay-0.7
- The Makefile will use the git version, or when that is not available, use the directory name to build gitversion .h
- Describe how this works in the README and include a link to the latest release archive.
- Someone that uses the github tip zipfile will get the gitversion.h as usbrelay-master. I've not had anyone using the zipfiles before so this is rare.

darrylb123 commented:
Michal,
Just read the story on LWN. Very cool stuff. Recognised the relays in the picture :)

wentasah commented:
Yes, this is it :) And under the table is a rack full of other pieces of hardware and relays.

Regrding this PR - I want to investigate a bit how github releases work and how we can automate creation of archives with `gitversion.c`. So far I had no time for playing with it, but it seems that https://github.com/actions/upload-release-asset could be what's needed.

wentasah commented:
It's long time since I created this PR. I thought, that it will be possible to force GitHub to generate release archives differently than how it's done by default, but it seems impossible. So I changed the approach.

Instead of creating the archives, now `make release` creates a commit with `version.mk` file added and then the file is removed in the next commit. This way, github releases will always contain correct version number in that file and we will not need to "guess" it from the directory name.

In addition to adding a version number, I added a `sign-release` target. It signs the release with GPG. It was [pointed out by a Debian developer](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1010494#10) that this would be nice thing to have.

@darrylb123 What do you think? Do you want to create releases by running `make release`? You can try it locally and see what it does; it performs all things locally, without touching github.

If you have a GPG key ready, you can also try signing the old releases. For example, running `make sign-release VER=1.0` will sign the last release. It should be no problem that the release was not created by `make release` from this commit. **Signing modifies the release on github** (it adds a signature file). It uses the `gh` tool (which must be installed) to communicate with github. If you want to try signing safely, I suggest adding `-n` the command line, which just prints the commands without executing them and you can execute them manually.

darrylb123 commented:
I notice the Makefile is not for the python wheel format now in master.
In the process of converting to wheel format I moved all the python files to a separate directory (usbrelay_py) and removed the python building from the Makefile and added it's own.
Now you cd to usbrelay_py and just type make.

This stuffed Mark around too. Perhaps make python should still be in the top level Makefile which just runs make -C usbrelay_py...

Mark was prepared to live with the change though. I don't mind either way.

wentasah commented:
Yes, I'd try to keep the interface for users the same. Toplevel `make python` would invoke `$(MAKE) -C usbrelay_py` (using `$(MAKE)` instead of `make` is better because when you run `make --dry-run`, the former does not execute the sub-make, while the latter does).

Don't you want to create a PR for python_wheel branch so that we can discuss it there and see the code in question?

darrylb123 commented:
Michal,
Sorry, I'm not at all an expert with git.
I created a python_wheel branch to change over to the wheel format and
incorporate the structure required by Fedora for python code. I was guided
through this by Mark.
There was a delay before it was used and I ended up with both master and
python_wheel diverging. (My fault), I'd cloned again and forgot to check
out python wheel...
I couldn't merge python_wheel because of the divergence, so I manually
merged the changes from python_wheel back into master.
I imagine there is a correct way to do the merge that would keep the
metadata correct.

I changed the documentation to the make in each directory, but I'm happy to
make the top level Makefile accept make python.

Darryl



On Sat, May 7, 2022 at 1:15 AM Michal Sojka ***@***.***>
wrote:

> Yes, I'd try to keep the interface for users the same. Toplevel make
> python would invoke $(MAKE) -C usbrelay_py (using $(MAKE) instead of make
> is better because when you run make --dry-run, the former does not
> execute the sub-make, while the latter does).
>
> Don't you want to create a PR for python_wheel branch so that we can
> discuss it there and see the code in question?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/darrylb123/usbrelay/pull/55#issuecomment-1119728729>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTSUVNVHYC5TKJAO3ETGQTVIUZQDANCNFSM4ND4GOCA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

darrylb123 commented:
Shall I leave it to you to add the make python and make install_py to the
top level Makefile while you are merging your make release changes?

Darryl


On Sat, May 7, 2022 at 1:27 PM Darryl Bond ***@***.***> wrote:

> Michal,
> Sorry, I'm not at all an expert with git.
> I created a python_wheel branch to change over to the wheel format and
> incorporate the structure required by Fedora for python code. I was guided
> through this by Mark.
> There was a delay before it was used and I ended up with both master and
> python_wheel diverging. (My fault), I'd cloned again and forgot to check
> out python wheel...
> I couldn't merge python_wheel because of the divergence, so I manually
> merged the changes from python_wheel back into master.
> I imagine there is a correct way to do the merge that would keep the
> metadata correct.
>
> I changed the documentation to the make in each directory, but I'm happy
> to make the top level Makefile accept make python.
>
> Darryl
>
>
>
> On Sat, May 7, 2022 at 1:15 AM Michal Sojka ***@***.***>
> wrote:
>
>> Yes, I'd try to keep the interface for users the same. Toplevel make
>> python would invoke $(MAKE) -C usbrelay_py (using $(MAKE) instead of make
>> is better because when you run make --dry-run, the former does not
>> execute the sub-make, while the latter does).
>>
>> Don't you want to create a PR for python_wheel branch so that we can
>> discuss it there and see the code in question?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/darrylb123/usbrelay/pull/55#issuecomment-1119728729>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABTSUVNVHYC5TKJAO3ETGQTVIUZQDANCNFSM4ND4GOCA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

darrylb123 commented:
I've updated the top level Makefile to make it backwards compatible and added an install_all.
Now make python
   -invokes make on the usbrelay_py
make install_py 
   -similar
make all 
   -builds the top level and usbrelay_py
make install_all 
   -installs everything

wentasah commented:
OK, I'll resolve the conflict. 

I'm not sure that add `python` to `all` is a good thing. Currently, on my system, it fails with:
```
make -C usbrelay_py
make[1]: Entering directory '/home/wsh/src/usbrelay/usbrelay_py'
python3 -m build --wheel
/usr/bin/python3: No module named build
```
This could confuse people who just want the `usbrelay` command.

wentasah commented:
I rebased the branch and removed the conflict.

darrylb123 commented:
Ok, I'll pull that and then remove python from all.

On Mon, 9 May 2022, 8:31 pm Michal Sojka, ***@***.***> wrote:

> I rebased the branch and removed the conflict.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/darrylb123/usbrelay/pull/55#issuecomment-1120927601>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTSUVPWXYLU7ARJTIZVWKTVJDSP3ANCNFSM4ND4GOCA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

darrylb123 commented:
Removed python from all and removed install_all.
Back to old behaviour
make 
sudo make install
make python
sudo make install_py

darrylb123 commented:
Michal,
In the process of helping with the RPM spec file I had to reference the tar file for a particular release sanely. 
I'd forgotten about your work on this.
I used the GitHub release process through the web site. It generates tar.gz and zip file that can be referenced in a consistent manner. A snippet from the spec file:
Name:          usbrelay
Version:       1.1
Release:       %autorelease
Summary:       A library and command line tool for controlling USB-connected relays based on hidapi
License:       GPL-2.0-or-later
URL:           https://github.com/darrylb123/%{name}
Source0:       %{url}/archive/%{version}/%{name}-%{version}.tar.gz

So, after creating a release (eg 1.1), the URL is
https://github.com/darrylb123/usbrelay/archive/1.1/usbrelay-1.1.tar.gz
